### PR TITLE
VM Tests: Run customized OS in VM.

### PR DESCRIPTION
**Pull request by cwize1**
Expand the basic test so that it takes the image produced by image customizer and then runs it in a VM using QEMU/KVM via libvirt.

---

### **Checklist**
- [x] Tests added/updated
- [x] Documentation updated (if needed)
- [x] Code conforms to style guidelines

### CODE REVIEW

File: test/vmtests/vmtests/test_no_change.py [hunk 1]
```diff
@@ -26,3 +37,25 @@ def test_no_change(
...
+    domain_xml = create_libvirt_domain_xml(VmSpec(vm_name, 4096, 4, diff_image_path))
+
+    vm = LibvirtVm(vm_name, domain_xml, libvirt_conn)
+    close_list.append(vm)
+
+    # Start VM.
+    vm.start()
+
+    # Wait for VM to boot by waiting for it to request an IP address from the DHCP server.
+    vm.get_vm_ip_address(timeout=15)
```
Comment (gmileka) [inline test/vmtests/vmtests/test_no_change.py:61]: 15 seconds might be too short? (my local tests usually take longer)

Comment (gmileka) [reply test/vmtests/vmtests/test_no_change.py:61]: but it's for iso, so, there are a few more things that take place...

File: test/vmtests/vmtests/utils/libvirt_utils.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+def _gen_disk_device_name(prefix: str, next_disk_indexes: Dict[str, int]) -> str:
+    disk_index = next_disk_indexes.get(prefix, 0)
+    next_disk_indexes[prefix] = disk_index + 1
+
+    match prefix:
+        case "vd" | "sd":
+            # The disk device name is required to follow the standard Linux device naming
+            # scheme. That is: [ sda, sdb, ..., sdz, sdaa, sdab, ... ]. However, it is
+            # unlikely that someone will ever need more than 26 disks. So, keep is simple
```
Comment (gmileka) [inline test/vmtests/vmtests/utils/libvirt_utils.py:148]: nit: `keep it`

File: test/vmtests/vmtests/utils/libvirt_vm.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        interface_name = next(iter(interfaces))
+        addrs = interfaces[interface_name]["addrs"]
+        if len(addrs) < 1:
+            return None
+
+        addr = addrs[0]["addr"]
+        assert isinstance(addr, str)
+        return addr
+
+    def close(self) -> None:
```
Comment (gmileka) [inline test/vmtests/vmtests/utils/libvirt_vm.py:62]: nit: would it be more clear to have two separate funcitons: `stop` and `delete`? The `close` can be the destructor calling both `stop` and `delete`.

File: test/vmtests/vmtests/test_no_change.py (hunk 1 above)
Comment (cwize1) [reply test/vmtests/vmtests/test_no_change.py:61]: In my local testing, I see 6-11 seconds. So, I'll bump this up to 30 seconds.

File: test/vmtests/vmtests/utils/local_client.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                # Log the line.
+                if log_enabled:
+                    line_strip_newline = line[:-1] if line.endswith("\n") else line
+                    logging.log(self._log_level, "%s: %s", self._log_name, line_strip_newline)
+
+            self._pipe.close()
+            self._output = output.getvalue()
+
+
+class LocalProcess:
```
Comment (gmileka) [inline test/vmtests/vmtests/utils/local_client.py:92]: Are we writing a new framework for running processes or is it being ported? Is this related to LISA in anyway? (I don't have context - but probably we have similar code somewhere)

Comment (cwize1) [reply test/vmtests/vmtests/utils/local_client.py:92]: This is new code. But the API was inspired by the [LISA node](https://github.com/microsoft/lisa/blob/main/lisa/node.py#L244) type. Though, I did simplify some of the params. 

Having stderr and stdout be automatically logged while also collecting them as strings is incredibly useful. It makes debugging test failures a lot easier. Hence, I copied the idea.

In a future change, I will add a very similar API for SSH.

File: test/vmtests/vmtests/utils/libvirt_vm.py (hunk 3 above)
Comment (cwize1) [reply test/vmtests/vmtests/utils/libvirt_vm.py:62]: There are actually 2 stop methods that libvirt exposes: graceful and default. So, if were to add a `stop` method, should I expose both or only one of them?

In addition, what should be the exception behavior of the `stop` and `delete` functions? Should they be best effort (like `close` is) or should they fail fast on the first error that is hit?

To be honest, I am not sure this test suite will ever need the ability to `stop` the VM outside of cleanup. So, I am inclined to keep it simple for now. If we ever need the extra functions, we can deal with the complexity then.

File: test/vmtests/vmtests/utils/libvirt_vm.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        # self.console_logger.attach(domain, console_log_file_path)
+
+        # Start the VM.
+        self.domain.resume()
+
+    # Wait for the VM to boot and then get the IP address.
+    def get_vm_ip_address(self, timeout: float = 30) -> str:
+        timeout_time = time.time() + timeout
+
+        while True:
```
[the remark this answers is not in the document]
Comment (cwize1) [reply test/vmtests/vmtests/utils/libvirt_vm.py:37]: Fixed.

File: test/vmtests/vmtests/utils/libvirt_utils.py (hunk 2 above)
Comment (cwize1) [reply test/vmtests/vmtests/utils/libvirt_utils.py]: Fixed.

File: test/vmtests/vmtests/test_no_change.py (hunk 1 above)
Comment (cwize1) [reply test/vmtests/vmtests/test_no_change.py]: Fixed.

**Review by gmileka — APPROVED** [review]
